File: src/main/java/no/nav/bidrag/beregn/forholdsmessigfordeling/beregning/ForholdsmessigFordelingBeregningImpl.java
package no.nav.bidrag.beregn.forholdsmessigfordeling.beregning;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import no.nav.bidrag.beregn.felles.enums.ResultatKode;
import no.nav.bidrag.beregn.forholdsmessigfordeling.bo.BeregnetBidragSak;
import no.nav.bidrag.beregn.forholdsmessigfordeling.bo.GrunnlagBeregningPeriodisert;
import no.nav.bidrag.beregn.forholdsmessigfordeling.bo.GrunnlagPerBarn;
import no.nav.bidrag.beregn.forholdsmessigfordeling.bo.ResultatBeregning;
import no.nav.bidrag.beregn.forholdsmessigfordeling.bo.ResultatPerBarn;

public class ForholdsmessigFordelingBeregningImpl implements ForholdsmessigFordelingBeregning {

  @Override
  public List<ResultatBeregning> beregn(
      GrunnlagBeregningPeriodisert grunnlagBeregningPeriodisert) {

    var resultatBeregningListe = new ArrayList<ResultatBeregning>();
    var resultatPerBarnListe = new ArrayList<ResultatPerBarn>();

    var endeligBidragsevne = BigDecimal.ZERO;
    if (grunnlagBeregningPeriodisert.getBidragsevne().getBidragsevneBelop().compareTo(
        grunnlagBeregningPeriodisert.getBidragsevne().getTjuefemProsentInntekt()) > 0){
      endeligBidragsevne = grunnlagBeregningPeriodisert.getBidragsevne().getBidragsevneBelop();
      } else {
      endeligBidragsevne = grunnlagBeregningPeriodisert.getBidragsevne().getTjuefemProsentInntekt();
    }

    var resultatkode = ResultatKode.KOSTNADSBEREGNET_BIDRAG;

    // Summerer beløp for alle saker i perioden. Hvis dette beløpet er høyere enn bidragsevnen så
    // skal det gjøres en forholdsmessig fordeling
    var samletBidragsbelopAlleSaker = grunnlagBeregningPeriodisert.getBeregnetBidragSakListe()
        .stream()
        .map(BeregnetBidragSak::getGrunnlagPerBarnListe)
        .flatMap(Collection::stream)
//        .collect(Collectors.toList())
//        .stream()
        .map(GrunnlagPerBarn::getBidragBelop)
//        .flatMap(GrunnlagPerBarn::getBidragBelop)
        .reduce(BigDecimal.ZERO, BigDecimal::add);
    System.out.println("Samlet bidragsbeløp for alle saker: " + samletBidragsbelopAlleSaker);

    if (samletBidragsbelopAlleSaker.compareTo(endeligBidragsevne) > 0){
      // Gjør forholdsmessig fordeling
      var sakensAndelAvTotaltBelopAlleSaker = BigDecimal.ZERO;
      var barnetsAndelAvTotaltBelopForSak = BigDecimal.ZERO;
      var beregnetBidragsbelop = BigDecimal.ZERO;
      var samletBidragsbelopSak = BigDecimal.ZERO;

      for (BeregnetBidragSak beregnetBidragSak : grunnlagBeregningPeriodisert.getBeregnetBidragSakListe()) {
        samletBidragsbelopSak = grunnlagBeregningPeriodisert.getBeregnetBidragSakListe()
            .stream()
            .filter(i -> i.getSaksnr() == beregnetBidragSak.getSaksnr())
            .map(BeregnetBidragSak::getGrunnlagPerBarnListe)
            .flatMap(Collection::stream)
            .map(GrunnlagPerBarn::getBidragBelop)
            .reduce(BigDecimal.ZERO, BigDecimal::add);
        System.out.println("Samlet bidragsbeløp for sak: " + beregnetBidragSak.getSaksnr() + " = " + samletBidragsbelopSak);






      }




      } else {
      // Ingen forholdsmessig fordeling gjøres og originalt bidragsbeløp returneres
      for (BeregnetBidragSak beregnetBidragSak : grunnlagBeregningPeriodisert.getBeregnetBidragSakListe()) {
        for (GrunnlagPerBarn grunnlagPerBarn : beregnetBidragSak.getGrunnlagPerBarnListe()){
          resultatPerBarnListe.add(new ResultatPerBarn(grunnlagPerBarn.getBarnPersonId(),
              grunnlagPerBarn.getBidragBelop(), ResultatKode.KOSTNADSBEREGNET_BIDRAG));
        }
        resultatBeregningListe
            .add(new ResultatBeregning(
                beregnetBidragSak.getSaksnr(),
                resultatPerBarnListe));


        var tempBarnebidrag = BigDecimal.ZERO;

      }


    }


    return resultatBeregningListe;
  }


}

File: src/main/kotlin/no/nav/bidrag/beregn/forholdsmessigfordeling/beregning/ForholdsmessigFordelingBeregningImpl.kt
package no.nav.bidrag.beregn.forholdsmessigfordeling.beregning

import no.nav.bidrag.beregn.forholdsmessigfordeling.bo.GrunnlagBeregningPeriodisert
import no.nav.bidrag.beregn.forholdsmessigfordeling.bo.ResultatBeregning
import no.nav.bidrag.beregn.forholdsmessigfordeling.bo.ResultatPerBarn
import no.nav.bidrag.domene.enums.beregning.ResultatkodeBarnebidrag
import java.math.BigDecimal
import java.math.MathContext
import java.math.RoundingMode

class ForholdsmessigFordelingBeregningImpl : ForholdsmessigFordelingBeregning {
    override fun beregn(grunnlag: GrunnlagBeregningPeriodisert): List<ResultatBeregning> {
        val resultatBeregningListe = mutableListOf<ResultatBeregning>()

        val endeligBidragsevne = minOf(grunnlag.bidragsevne.belop, grunnlag.bidragsevne.tjuefemProsentInntekt)

        // Hvis summert beløp for alle saker i perioden er høyere enn bidragsevnen skal det gjøres en forholdsmessig fordeling
        val samletBidragsbelopAlleSaker =
            grunnlag.beregnetBidragSakListe
                .flatMap { it.grunnlagPerBarnListe }
                .map { it.bidragBelop }
                .fold(BigDecimal.ZERO, BigDecimal::add)

        // Gjør forholdsmessig fordeling
        if (samletBidragsbelopAlleSaker > endeligBidragsevne) {
            grunnlag.beregnetBidragSakListe.forEach { beregnetBidragSak ->
                val samletBidragsbelopSak =
                    grunnlag.beregnetBidragSakListe
                        .filter { it.saksnr == beregnetBidragSak.saksnr }
                        .flatMap { it.grunnlagPerBarnListe }
                        .map { it.bidragBelop }
                        .fold(BigDecimal.ZERO, BigDecimal::add)

                // Hvor stor prosentdel av det totale beløpet utgjør denne sakens samlede bidragsbeløp
                val sakensAndelAvTotaltBelopAlleSakerProsent =
                    samletBidragsbelopSak.divide(
                        samletBidragsbelopAlleSaker,
                        MathContext(10, RoundingMode.HALF_UP),
                    )

                // Regner ut hvor mye prosentsatsen utgjør av bidragsevnen
                val sakensAndelAvEndeligBidragsevneBelop =
                    sakensAndelAvTotaltBelopAlleSakerProsent.multiply(
                        endeligBidragsevne,
                        MathContext(10, RoundingMode.HALF_UP),
                    )

                val resultatPerBarnListe = mutableListOf<ResultatPerBarn>()

                // Leser hvert barn i saken og regner ut nytt, justert bidragsbeløp
                beregnetBidragSak.grunnlagPerBarnListe.forEach { grunnlagPerBarn ->
                    val barnetsAndelAvTotaltBelopForSakProsent =
                        grunnlagPerBarn.bidragBelop.divide(
                            samletBidragsbelopSak,
                            MathContext(10, RoundingMode.HALF_UP),
                        )
                    val beregnetBidragsbelop =
                        barnetsAndelAvTotaltBelopForSakProsent.multiply(
                            sakensAndelAvEndeligBidragsevneBelop,
                            MathContext(10, RoundingMode.HALF_UP),
                        ).setScale(-1, RoundingMode.HALF_UP)

                    resultatPerBarnListe.add(
                        ResultatPerBarn(
                            barnPersonId = grunnlagPerBarn.barnPersonId,
                            belop = beregnetBidragsbelop,
                            kode = ResultatkodeBarnebidrag.FORHOLDSMESSIG_FORDELING_BIDRAGSBELØP_ENDRET,
                        ),
                    )
                }

                resultatBeregningListe.add(
                    ResultatBeregning(saksnr = beregnetBidragSak.saksnr, resultatPerBarnListe = resultatPerBarnListe),
                )
            }

            // Ingen forholdsmessig fordeling gjøres og originalt bidragsbeløp returneres
        } else {
            grunnlag.beregnetBidragSakListe.forEach {
                val resultatPerBarnListe = mutableListOf<ResultatPerBarn>()

                it.grunnlagPerBarnListe.forEach { grunnlagPerBarn ->
                    resultatPerBarnListe.add(
                        ResultatPerBarn(
                            barnPersonId = grunnlagPerBarn.barnPersonId,
                            belop = grunnlagPerBarn.bidragBelop,
                            kode = ResultatkodeBarnebidrag.FORHOLDSMESSIG_FORDELING_INGEN_ENDRING,
                        ),
                    )
                }

                resultatBeregningListe.add(ResultatBeregning(saksnr = it.saksnr, resultatPerBarnListe = resultatPerBarnListe))
            }
        }

        return resultatBeregningListe
    }
}
